File: problems/0901-1000/0901_online-stock-span/solution.cpp
#include <utility>
#include <vector>

class StockSpanner {
  public:
    StockSpanner() = default;

    int next(int price) {
        int span = 1;
        while (!stack.empty() && stack.back().first <= price) {
            span += stack.back().second;
            stack.pop_back();
        }
        stack.emplace_back(price, span);
        return span;
    }

  private:
    std::vector<std::pair<int, int>> stack;
};

File: problems/0901-1000/0901_online-stock-span/solution.rs
pub struct StockSpanner {
    stack: Vec<(i32, i32)>,
}

impl StockSpanner {
    pub fn new() -> Self {
        StockSpanner { stack: Vec::new() }
    }

    pub fn next(&mut self, price: i32) -> i32 {
        let mut span = 1;
        while let Some(&(top, top_span)) = self.stack.last() {
            if top > price {
                break;
            }
            span += top_span;
            self.stack.pop();
        }
        self.stack.push((price, span));
        span
    }
}
